File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/runtime/TransactionalUnitOfWork.java
package com.zappymart.adapters.runtime;

import com.zappymart.application.ports.UnitOfWork;
import org.springframework.transaction.support.TransactionTemplate;

import java.util.function.Supplier;

public final class TransactionalUnitOfWork implements UnitOfWork {

    private final TransactionTemplate transactionTemplate;

    public TransactionalUnitOfWork(TransactionTemplate transactionTemplate) {
        this.transactionTemplate = transactionTemplate;
    }

    @Override
    public <TValue> TValue inTransaction(Supplier<TValue> work) {
        return transactionTemplate.execute(status -> work.get());
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/transaction/TransactionalUnitOfWork.kt
package nl.zappymart.adapters.transaction

import nl.zappymart.application.ports.UnitOfWork
import org.springframework.stereotype.Component
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.support.TransactionTemplate

@Component
class TransactionalUnitOfWork(transactionManager: PlatformTransactionManager) : UnitOfWork {

    private val template = TransactionTemplate(transactionManager)

    override fun <Value> execute(work: () -> Value): Value =
        requireNotNull(template.execute { _ -> Outcome(work()) }).value

    private class Outcome<Value>(val value: Value)
}
